### PR TITLE
JP2OpenJPEG: refactor driver code into re-usable template

### PR DESCRIPTION
**Pull request by boxerab**
Once this PR is validated, it will be re-used in
https://github.com/OSGeo/gdal/pull/8133

## What does this PR do?

This PR is related to https://github.com/OSGeo/gdal/pull/8133
It will validate the changes made in that PR and provide a reusable template
for both OpenJPEG and Grok JPEG 2000 drivers.

## Tasklist

 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: frmts/openjpeg/opjcontext.h [hunk 1]
```diff
@@ -0,0 +1,1189 @@
...
+    jp2_image_comp_param *pasBandParams;
+};
+
+/************************************************************************/
+/* ==================================================================== */
+/*                           OPJContext                                 */
+/* ==================================================================== */
+/************************************************************************/
+
+struct OPJContext : public JP2Context
```
[the remark this answers is not in the document]
Comment (boxerab) [reply frmts/openjpeg/opjcontext.h]: Thanks, it **is** confusing :) I've renamed things - local context is actually a wrapper around low level codec functionality, and context is actually the base class for the dataset.

File: frmts/opjlike/jp2opjlikedataset.cpp [hunk 2]
```diff
@@ -0,0 +1,3745 @@
+/******************************************************************************
+ *
+ * Project:  JPEG2000 driver based on OpenJPEG or Grok library
+ * Purpose:  JPEG2000 driver based on OpenJPEG or Grok library
+ * Author:   Aaron Boxer, <boxerab at protonmail dot com>
```
Comment (rouault) [inline frmts/opjlike/jp2opjlikedataset.cpp:5]: I wouldn't mind being kept as one of the co-authors:
```suggestion
 * Authors:   Even Rouault, <even dot rouault at spatialys dot com>
 *            Aaron Boxer, <boxerab at protonmail dot com>
```

**Comment by rouault** [comment]
I believe we must be close to be good to go (with my above remark). Can you git rebase on top of latest master to just keep the relevant commit ?

File: frmts/opjlike/jp2opjlikedataset.cpp (hunk 2 above)
Comment (boxerab) [reply frmts/opjlike/jp2opjlikedataset.cpp]: Good idea, done.

**Comment by boxerab** [comment]
> I believe we must be close to be good to go (with my above remark). Can you git rebase on top of latest master to just keep the relevant commit ?

Thanks, rebased and ready to merge

**Comment by rouault** [comment]
Merging

**Pull request merged**